sql: use SQLInstanceID in EXPLAIN instead of OptionalNodeIDErr

asubiotto opened the pull request:
Tenants would previously not be able to run EXPLAIN because they wouldn't have
access to a NodeID. This is now fixed by using the SQLInstanceID instead, which
will provide an expected SQL-scoped identifier in a multi-tenant environment
and will translate to a NodeID in a single-tenant environment.

Release note: None (multitenancy-related change)

Closes #49072

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49585)
<!-- Reviewable:end -->

tbg commented on pkg/sql/explain_distsql.go:
```diff
@@ -227,11 +227,7 @@ func (n *explainDistSQLNode) startExec(params runParams) error {
-			return err
-		}
-		flows := plan.GenerateFlowSpecs(nodeID)
+		flows := plan.GenerateFlowSpecs(roachpb.NodeID(params.extendedEvalCtx.NodeID.SQLInstanceID()))
```
Casting to NodeID makes me uneasy on general principles as confusing them in general can lead to correctness problems. Can you get away with a zero NodeID here and elsewhere? Taking a step back, SQL flows for a tenant will never be scheduled on a "Node" in a multi-tenant world, it will always be a tenant. I don't think we necessarily want to do the code changes for that now (since we won't schedule nonlocal flows anyway, at least not for the time being), but it seems good to keep the concepts as separate as we can until then.

asubiotto replied:
I agree. the best way to go about this is to change `GenerateFlowSpecs` to work with SQL instance IDs, but that'll mean changing a lot of the planning code which doesn't seem worth to do at this stage. I'm not sure whether a 0 node ID would work.

My goal was to just get `EXPLAIN` working. I think having this cast is worth the cost of making that work. I should at least add a TODO here though. Do you prefer to hold off on merging this?

tbg replied:
No, this looks like a SQL team decision and that's you :-) Go ahead and merge.

asubiotto replied:
Added a TODO and a linked issue.

Commit:
sql: use SQLInstanceID in EXPLAIN instead of OptionalNodeIDErr

Tenants would previously not be able to run EXPLAIN because they wouldn't have
access to a NodeID. This is now fixed by using the SQLInstanceID instead, which
will provide an expected SQL-scoped identifier in a multi-tenant environment
and will translate to a NodeID in a single-tenant environment.

Release note: None (multitenancy-related change)

asubiotto commented:
bors r=tbg
